Keep filters option when switching repos

Arif-Khalid opened the pull request:
### Summary:

Fixes #205 

#### Type of change:

- ✨ New Feature/ Enhancement

### Changes Made:

- Added checkbox on dialog that opens when selecting a repo
- If that checkbox is checked when confirming selection, filter selection for status, type, sort, search bar, selected labels and hidden labels are kept
- Checkbox is not mounted when there is no repo currently selected, i.e., no changes to keep

### Screenshots:
* Before the change
<img width="372" alt="image" src="https://github.com/CATcher-org/WATcher/assets/88131400/f076258f-4a2c-4f90-b830-05ec01d26439">

* After the change
<img width="373" alt="image" src="https://github.com/CATcher-org/WATcher/assets/88131400/f5026d3e-8344-437f-aea3-98550c7b1c4c">


### Proposed Commit Message:

```
Option to keep filters you have set on repo changes.

No way to keep filters when changing repos.

An option to keep the common filters will be useful for anyone
managing multiple repositories without the hassle of reapplying 
the same filters.

Let's add a checkbox to the change repo dialog allowing 
users to keep their filters.
```

<details><summary>
<h3>Checklist:</h3>
</summary>

- [x] I have tested my changes thoroughly.
- [ ] I have created tests for any new code files created in this PR or provided a link to a issue/PR that addresses this.
- [x] I have added or modified code comments to improve code readability where necessary.
- [ ] I have updated the project's documentation as necessary.

</details>

gycgabriel reviewed (changes requested):
Tested - The filters are shown but they are not applied after switching the repo
Also, the search bar and the label filters are not persisted.

Arif-Khalid commented on src/app/shared/issue-tables/dropdownfilter.ts:
```diff
@@ -18,7 +18,43 @@ export const DEFAULT_DROPDOWN_FILTER = <DropdownFilter>{
+let currentDropdownFilter: DropdownFilter = { ...DEFAULT_DROPDOWN_FILTER };
+
+/** This is the initial dropdown filter applied when first viewing a new repo */
+let initialDropdownFilter: DropdownFilter = { ...DEFAULT_DROPDOWN_FILTER };
```
Both are unnecessary in hindsight. Initial dropdown filter variable has been removed.

Arif-Khalid commented:
> Tested - The filters are shown but they are not applied after switching the repo Also, the search bar and the label filters are not persisted.

Committed new changes that fixes functionality for status, type and sort filters.
Labels and issue names are likely always different across repos. Should I continue to work on storing search bar and labels?
If so should I ignore labels that don't exist on the new repo?

gycgabriel commented:
> Labels and issue names are likely always different across repos. Should I continue to work on storing search bar and labels?

Actually, for CS2103 projects which is the main use case for WATcher, they will most likely be same or similar. It's up to you if you want to continue working on it or create a separate issue for someone else to pick up. 


> If so should I ignore labels that don't exist on the new repo?

Yes, drop label selections that don't exist

Arif-Khalid commented:
> > Labels and issue names are likely always different across repos. Should I continue to work on storing search bar and labels?
> 
> Actually, for CS2103 projects which is the main use case for WATcher, they will most likely be same or similar. It's up to you if you want to continue working on it or create a separate issue for someone else to pick up.
> 
> > If so should I ignore labels that don't exist on the new repo?
> 
> Yes, drop label selections that don't exist

I have refactored the code into its own file called FiltersStore.ts and am using a static class to store the filters as they are edited, initialize with previous filter values and clearing them if filters are not kept.
Current status is that functionality for search, status, type and sort filters are working; Also fixed an oversight that ascending and descending order of sort was not accounted for.
Before I start work on the labels, I would appreciate some feedback on the direction and I have taken and if necessary a suggestion for a better, alternative approach.

Arif-Khalid commented:
In hindsight, I think a better approach for consistency would be to have my FiltersStore.ts be the central home of the filters such that you update and retrieve from FiltersStore. This would negate me having to copy filter changes and will be more developer friendly to understand. What do you think?

Arif-Khalid commented on src/app/shared/filter-bar/filter-bar.component.ts:
```diff
@@ -24,7 +25,10 @@ export class FilterBarComponent implements OnInit, AfterViewInit, OnDestroy {
 
   /** Selected dropdown filter value */
-  dropdownFilter: DropdownFilter = DEFAULT_DROPDOWN_FILTER;
+  dropdownFilter: DropdownFilter = FiltersStore.getInitialDropdownFilter();
+
+  /** The initial search filter value, not updated when search filter is changed */
+  initialSearchFilter: string = FiltersStore.getInitialSearchFilter();
 
```
It is named initial because this search filter variable is not updated at all when the user types. Instead this variable is simply used to initialize what is seen by the user at the search bar when they first open a new repo.

Arif-Khalid commented:
I have updated functionality to include the working logic for storing the status, type, sort, search bar, selected labels and hidden labels when the keep filters option is checked when switching repos. I also encountered a small bug where the DEFAULT_DROPDOWN_FILTER milestones property was being added to when switching repos since the array containing the milestones was passed by reference. My most recent push also fixes this issue.

gycgabriel reviewed (commented):
I'm getting errors for structuredClone, is this necessary 

ERROR in src/app/shared/issue-tables/FiltersStore.ts:8:58 - error TS2304: Cannot find name 'structuredClone'.

only available in node 17, we're using node 14

gycgabriel commented:
I'm not completely convinced that we need both initial and current variables, since by default it will reset, so why do we have to store the initial state?

Arif-Khalid commented:
Will close this PR to work on #249 as the fix for #205 will be easy after.
